File: languages/go/par-parallel-reduction/parallel_reduction.go
package main

import (
    "encoding/json"
    "fmt"
    "math"
    "os"
    "strconv"
    "sync"
)

type Result struct { ItemCount int `json:"item_count"`; ValueSum int `json:"value_sum"`; Workers int `json:"workers"` }
func transform(x int) int { return (x*7 + 11) % 1000 }
func main() {
    itemCount := 200000
    if len(os.Args) > 1 { if v, err := strconv.Atoi(os.Args[1]); err == nil { itemCount = v } }
    workers := 4
    chunk := int(math.Ceil(float64(itemCount) / float64(workers)))
    ch := make(chan [2]int, workers)
    var wg sync.WaitGroup
    for start := 0; start < itemCount; start += chunk {
        end := start + chunk; if end > itemCount { end = itemCount }
        wg.Add(1)
        go func(s, e int) {
            defer wg.Done(); count, total := 0, 0
            for i := s; i < e; i++ { total += transform(i); count++ }
            ch <- [2]int{count, total}
        }(start, end)
    }
    go func(){ wg.Wait(); close(ch) }()
    totalCount, totalSum := 0, 0
    for r := range ch { totalCount += r[0]; totalSum += r[1] }
    out, _ := json.Marshal(Result{ItemCount: totalCount, ValueSum: totalSum, Workers: workers})
    fmt.Println(string(out))
}

File: languages/java/par-parallel-reduction/ParallelReduction.java
import java.util.ArrayList;
import java.util.List;

public class ParallelReduction {
    static int transform(int x) { return (x * 7 + 11) % 1000; }
    public static void main(String[] args) throws Exception {
        int itemCount = args.length > 0 ? Integer.parseInt(args[0]) : 200000;
        int workers = 4;
        int chunk = (int)Math.ceil((double)itemCount / workers);
        List<Thread> threads = new ArrayList<>();
        int[] counts = new int[workers];
        int[] sums = new int[workers];
        for (int w = 0; w < workers; w++) {
            final int idx = w;
            final int start = w * chunk;
            final int end = Math.min(start + chunk, itemCount);
            threads.add(new Thread(() -> {
                int c = 0; int s = 0;
                for (int i = start; i < end; i++) { s += transform(i); c++; }
                counts[idx] = c; sums[idx] = s;
            }));
        }
        for (Thread t : threads) t.start();
        for (Thread t : threads) t.join();
        int totalCount = 0; int totalSum = 0;
        for (int i = 0; i < workers; i++) { totalCount += counts[i]; totalSum += sums[i]; }
        System.out.println("{\"item_count\":" + totalCount + ",\"value_sum\":" + totalSum + ",\"workers\":" + workers + "}");
    }
}
